p2p/ipld/plugin/nodes/nodes.go

```diff
@@ -163,13 +148,91 @@ func DataSquareRowOrColumnRawInputParser(r io.Reader, _mhType uint64, _mhLen int
 	}
 	// to trigger the collection of nodes:
 	_ = n.Root()
-	return nodes, nil
+	return collector.ipldNodes(), nil
+}
+
+// nmtNodeCollector creates and collects ipld.Nodes if inserted into a nmt tree.
+// It is mainly used for testing.
+type nmtNodeCollector struct {
+	nodes []node.Node
+}
+
+func newNodeCollector() *nmtNodeCollector {
+	// The extendedRowOrColumnSize is hardcode this here to avoid importing:
+	// https://github.com/lazyledger/lazyledger-core/blob/585566317e519bbb6d35d149b7e856c4c1e8657c/types/consts.go#L23
+	const extendedRowOrColumnSize = 2 * 128
+	return &nmtNodeCollector{nodes: make([]node.Node, 0, extendedRowOrColumnSize)}
+}
+
+func (n nmtNodeCollector) ipldNodes() []node.Node {
+	return n.nodes
+}
+
+func (n *nmtNodeCollector) visit(hash []byte, children ...[]byte) {
+	cid := mustCidFromNamespacedSha256(hash)
+	switch len(children) {
+	case 1:
+		n.nodes = prependNode(nmtLeafNode{
+			cid:  cid,
+			Data: children[0],
+		}, n.nodes)
+	case 2:
+		n.nodes = prependNode(nmtNode{
+			cid: cid,
+			l:   children[0],
+			r:   children[1],
+		}, n.nodes)
+	default:
+		panic("expected a binary tree")
+	}
+}
+
+func prependNode(newNode node.Node, nodes []node.Node) []node.Node {
+	nodes = append(nodes, node.Node(nil))
+	copy(nodes[1:], nodes)
+	nodes[0] = newNode
+	return nodes
+}
+
+// NmtNodeAdder adds ipld.Nodes to the underlying ipld.Batch if it is inserted
+// into an nmt tree
+type NmtNodeAdder struct {
+	batch *format.Batch
+	ctx   context.Context
+}
+
+// NewNmtNodeAdder returns a new NmtNodeAdder with the provided context and
+// batch. Note that the context provided should have a timeout
+func NewNmtNodeAdder(ctx context.Context, batch *format.Batch) *NmtNodeAdder {
+	return &NmtNodeAdder{
+		batch: batch,
+		ctx:   ctx,
+	}
+}
+
+// Visit can be inserted into an nmt tree to create ipld.Nodes while computing the root
+func (n *NmtNodeAdder) Visit(hash []byte, children ...[]byte) {
+	cid := mustCidFromNamespacedSha256(hash)
+	switch len(children) {
+	case 1:
+		n.batch.Add(n.ctx, nmtLeafNode{
+			cid:  cid,
+			Data: children[0],
+		})
+	case 2:
+		n.batch.Add(n.ctx, nmtNode{
+			cid: cid,
+			l:   children[0],
+			r:   children[1],
+		})
+	default:
+		panic("expected a binary tree")
```

Comment:
Not that this is public API, Im wondering, if we should change the visitor to potentially return an error instead 🤔  This will require changes in the nmt library too. Orthogonal to this PR though.